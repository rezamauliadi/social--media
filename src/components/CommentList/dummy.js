export default [
  {
    postId: 1,
    id: 1,
    name: "id labore",
    email: "Eliseo@gardner.biz",
    body: "laudantium "
  },
  {
    postId: 1,
    id: 2,
    name: "quo vero ",
    email: "Jayne_Kuhic@sydney.com",
    body: "est natus enim nihil est "
  },
  {
    postId: 1,
    id: 3,
    name: "odio adipisci",
    email: "Nikita@garfield.biz",
    body: "quia molestiae reprehenderit "
  }
];
