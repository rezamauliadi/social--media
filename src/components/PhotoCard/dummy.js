const albums = [
  {
    userId: 1,
    id: 1,
    title: "quidem molestiae enim"
  },
  {
    userId: 2,
    id: 11,
    title: "quidem molestiae enim"
  }
];
const photos = [
  {
    albumId: 1,
    id: 1,
    title: "accusamus beatae ad facilis cum similique qui sunt",
    url: "https://via.placeholder.com/600/92c952",
    thumbnailUrl: "https://via.placeholder.com/150/92c952"
  },
  {
    albumId: 2,
    id: 51,
    title: "non sunt voluptatem placeat consequuntur rem incidunt",
    url: "https://via.placeholder.com/600/8e973b",
    thumbnailUrl: "https://via.placeholder.com/150/8e973b"
  }
];

export default { albums, photos };
