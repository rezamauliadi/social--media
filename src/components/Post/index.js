import React, { Component } from "react";

import { Link } from "react-router-dom";
import {
  Button,
  Comment,
  Divider,
  Feed,
  Form,
  Image,
  Label,
  Loader,
  Message
} from "semantic-ui-react";
import PostComment from "src/components/PostComment";
import PostForm from "src/components/PostForm";

import getAvatar from "src/helpers/avatar-helper";

class Post extends Component {
  state = {
    id: 0,
    title: "",
    body: "",
    tempTitle: "",
    tempBody: "",
    comments: [],
    showComments: false,
    showUpdateForm: false,
    hasFetchComments: false,
    fetchingComments: false,
    deletingPost: false,
    updatingPost: false,
    invalidForm: false,
    opacity: "1",
    commentButtonText: "Show Comments"
  };

  changeValue = (_event, { name, value }) => {
    this.setState({ [name]: value });
  };

  updatePost = async () => {
    this.setState({ updatingPost: true, invalidForm: false });

    const { id, body, title } = this.state;
    const { userId } = this.props;

    if (!body || !title) {
      this.setState({ updatingPost: false, invalidForm: true });
      return;
    }

    if (id > 100) {
      // handle post id not found if exceed 100
      this.setState({ showUpdateForm: false, updatingPost: false });
      return;
    }

    const url = `https://jsonplaceholder.typicode.com/posts/${id}`;
    const response = await fetch(url, {
      method: "PUT",
      body: JSON.stringify({ id, title, body, userId }),
      headers: {
        "Content-type": "application/json; charset=UTF-8"
      }
    });
    const updatedPost = await response.json();
    this.props.onUpdatePost(updatedPost);
    this.setState({ showUpdateForm: false, updatingPost: false });
  };

  deletePost = async () => {
    this.setState({ deletingPost: true });
    const { id } = this.state;
    const url = `http://jsonplaceholder.typicode.com/posts/${id}`;
    const response = await fetch(url, { method: "DELETE" });

    if (response.status === 200) {
      this.setState({ opacity: "0", deletePost: false });
      setTimeout(() => {
        this.props.onDeletePost(id);
      }, 500);
    }
  };

  getComments = async id => {
    const url = `http://jsonplaceholder.typicode.com/comments?postId=${id}`;
    const response = await fetch(url);
    const data = await response.json();
    return data;
  };

  pushNewComment = newComment => {
    const newCommentList = [...this.state.comments];
    newCommentList.push(newComment);
    this.setState({ comments: newCommentList });
  };

  toggleComments = async () => {
    const { id, comments, hasFetchComments, showComments } = this.state;

    if (!showComments) {
      if (comments.length || (!comments.length && hasFetchComments)) {
        this.setState({
          showComments: true,
          commentButtonText: "Hide Comments"
        });
      } else {
        this.setState({ fetchingComments: true });
        const comments = await this.getComments(id);
        this.setState({
          comments,
          commentButtonText: "Hide Comments",
          fetchingComments: false,
          hasFetchComments: true,
          showComments: true
        });
      }
    } else {
      this.setState({
        showComments: false,
        commentButtonText: "Show Comments"
      });
    }
  };

  commentButtonContent = () => {
    if (this.state.fetchingComments) {
      return (
        <div>
          <Loader style={{ marginRight: "4px" }} size="mini" active inline />{" "}
          Loading...
        </div>
      );
    } else {
      return this.state.commentButtonText;
    }
  };

  commentForm = () => {
    return (
      <div style={{ marginTop: "18px" }}>
        <PostForm type="comments" onSubmit={this.pushNewComment} />
      </div>
    );
  };

  commentList = () => {
    const { comments, id } = this.state;
    return comments.map(comment => (
      <PostComment comment={comment} key={`${id}-${comment.id}`} />
    ));
  };

  renderComments = () => {
    const { comments, hasFetchComments, showComments } = this.state;

    if (comments.length && showComments) {
      return (
        <div>
          {this.commentList()}
          {this.commentForm()}
        </div>
      );
    } else if (!comments.length && hasFetchComments && showComments) {
      return (
        <div>
          <Label basic>No comment yet...</Label>
          {this.commentForm()}
        </div>
      );
    } else {
      return <div />;
    }
  };

  toggleUpdateForm = isOpen => {
    if (isOpen) {
      const { title, body } = this.state;
      this.setState({
        tempTitle: title,
        tempBody: body,
        showUpdateForm: isOpen
      });
    } else {
      const { tempTitle, tempBody } = this.state;
      this.setState({
        title: tempTitle,
        body: tempBody,
        showUpdateForm: isOpen
      });
    }
  };

  renderPostContent = () => {
    const {
      title,
      body,
      showUpdateForm,
      invalidForm,
      deletingPost,
      updatingPost
    } = this.state;

    if (showUpdateForm) {
      return (
        <Form warning={invalidForm}>
          <Form.Input
            placeholder="Give it a title!"
            name="title"
            value={title}
            onChange={this.changeValue}
          />
          <Form.TextArea
            placeholder="Write your post!"
            name="body"
            value={body}
            rows="5"
            onChange={this.changeValue}
          />
          <Message size="mini" warning>
            It will be great if you put some words in the post and give it a
            cool title.
          </Message>
          <Button
            size="tiny"
            icon="send"
            primary
            loading={updatingPost}
            disabled={deletingPost || updatingPost}
            onClick={() => this.updatePost()}
          />
          <Button
            size="tiny"
            icon="cancel"
            color="orange"
            disabled={deletingPost || updatingPost}
            onClick={() => this.toggleUpdateForm(false)}
          />
        </Form>
      );
    } else {
      return (
        <div>
          <div style={{ fontWeight: "bold", fontStyle: "italic" }}>{title}</div>
          {body}
        </div>
      );
    }
  };

  async componentDidMount() {
    const { post } = this.props;
    this.setState({
      id: post.id,
      title: post.title,
      body: post.body
    });
  }

  render() {
    const { deletingPost, updatingPost, opacity } = this.state;
    const { user } = this.props;

    return (
      <div style={{ opacity, transition: "opacity 500ms linear" }}>
        <Feed>
          <Feed.Event>
            <Feed.Label>
              <Image src={getAvatar(user.id)} />
            </Feed.Label>
            <Feed.Content>
              <Feed.Summary>
                <Feed.User as={Link} to={`users/${user.id}/posts`}>
                  {user.name}
                </Feed.User>{" "}
                posted!
                <Feed.Date>1 day ago</Feed.Date>
              </Feed.Summary>
              <Feed.Extra text>{this.renderPostContent()}</Feed.Extra>
            </Feed.Content>
          </Feed.Event>
        </Feed>
        <Button
          size="mini"
          style={{ marginLeft: "50px" }}
          onClick={() => this.toggleComments()}
        >
          {this.commentButtonContent()}
        </Button>
        <Button
          size="tiny"
          color="red"
          floated="right"
          icon="trash"
          loading={deletingPost}
          disabled={deletingPost || updatingPost}
          onClick={() => this.deletePost()}
        />
        <Button
          size="tiny"
          color="yellow"
          floated="right"
          icon="pencil"
          loading={updatingPost}
          disabled={deletingPost || updatingPost}
          onClick={() => this.toggleUpdateForm(true)}
        />

        <Comment.Group style={{ marginLeft: "60px" }}>
          {this.renderComments()}
        </Comment.Group>
        <Divider />
      </div>
    );
  }
}

export default Post;
