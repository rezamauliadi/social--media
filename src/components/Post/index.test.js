import React from "react";
import { shallow } from "enzyme";

import Post from "./index.js";
import dummy from "./dummy.js";
import API from "src/api";

describe("Component: Post", () => {
  const fnMock = jest.fn();
  const props = {
    post: dummy.post,
    user: dummy.user,
    onDeletePost: fnMock,
    onUpdatePost: fnMock
  };

  it("should match snapshot", () => {
    const component = shallow(<Post {...props} />);
    expect(component).toMatchSnapshot();
  });

  describe("button click", () => {
    it("should call openUpdateForm when update button clicked", () => {
      const component = shallow(<Post {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "openUpdateForm");
      component
        .find("Button")
        .at(1)
        .simulate("click");
      expect(instance.openUpdateForm).toHaveBeenCalled();
    });

    it("should call deletePost when delete button clicked", () => {
      const component = shallow(<Post {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "deletePost");
      jest.spyOn(instance.props, "onDeletePost");

      component
        .find("Button")
        .at(0)
        .simulate("click");
      expect(instance.deletePost).toHaveBeenCalled();
      setTimeout(() => {
        expect(instance.props.onDeletePost).toHaveBeenCalled();
      }, 500);
    });
  });

  describe("afterSubmitForm", () => {
    it("should call onUpdatePost", () => {
      const component = shallow(<Post {...props} />);
      const instance = component.instance();

      jest.spyOn(instance.props, "onUpdatePost");
      instance.afterSubmitForm();

      expect(instance.props.onUpdatePost).toHaveBeenCalled();
      expect(component.state("showUpdateForm")).toBe(false);
    });
  });

  describe("deletePost", () => {
    it("should call deletePost API", done => {
      const component = shallow(<Post {...props} />);
      const instance = component.instance();

      jest.spyOn(API, "deletePost").mockImplementation(() => 200);
      jest.spyOn(instance.props, "onDeletePost");
      jest.spyOn(global, "setTimeout");

      instance.deletePost();
      process.nextTick(() => {
        expect(component.state("opacity")).toBe("0");
        expect(component.state("deletingPost")).toBe(false);

        expect(global.setTimeout).toHaveBeenCalled();
        expect(instance.props.onDeletePost).toHaveBeenCalled();

        API.deletePost.mockClear();
        done();
      });
    });
  });

  describe("openUpdateForm", () => {
    it("should set showUpdateForm to true", () => {
      const component = shallow(<Post {...props} />);
      const instance = component.instance();

      instance.openUpdateForm();
      expect(component.state("showUpdateForm")).toBe(true);
    });
  });

  describe("closeUpdateForm", () => {
    it("should set showUpdateForm to false", () => {
      const component = shallow(<Post {...props} />);
      const instance = component.instance();

      instance.closeUpdateForm();
      expect(component.state("showUpdateForm")).toBe(false);
    });
  });
});
