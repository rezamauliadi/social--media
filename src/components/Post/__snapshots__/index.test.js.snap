// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Post should match snapshot 1`] = `
<div
  style={
    Object {
      "opacity": "1",
      "transition": "opacity 500ms linear",
    }
  }
>
  <Feed>
    <FeedEvent>
      <FeedLabel>
        <Image
          as="img"
          src="/images/daniel.jpg"
          ui={true}
        />
      </FeedLabel>
      <FeedContent>
        <FeedSummary>
          <FeedUser
            as={[Function]}
            to="users/1/posts"
          >
            Leanne Graham
          </FeedUser>
           
          posted!
          <FeedDate>
            1 day ago
          </FeedDate>
        </FeedSummary>
        <FeedExtra
          text={true}
        >
          <div>
            <div
              style={
                Object {
                  "fontStyle": "italic",
                  "fontWeight": "bold",
                }
              }
            >
              foo
            </div>
            bar
          </div>
        </FeedExtra>
      </FeedContent>
    </FeedEvent>
  </Feed>
  <Button
    as="button"
    color="red"
    disabled={false}
    floated="right"
    icon="trash"
    loading={false}
    onClick={[Function]}
    size="tiny"
    style={
      Object {
        "display": "block",
      }
    }
  />
  <Button
    as="button"
    color="yellow"
    disabled={false}
    floated="right"
    icon="pencil"
    onClick={[Function]}
    size="tiny"
    style={
      Object {
        "display": "block",
      }
    }
  />
  <CommentList
    postId={1}
  />
  <Divider />
</div>
`;
